### ISSUE TITLE
Named Dimensions

### ISSUE THREAD

**Issue by danijar** [tensorflow/tensorflow#38228]
**System information**

- TensorFlow version (you are using): 2.1
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**

Currently, TensorFlow ops receive axes as integer indices, e.g. `tf.reduce_sum(tensor, axis=(2, 3))`. Keep track of axes after applying several operations on them and passing tensors around through multiple functions can be challenging.

I've seen many researchers and developers grab a sheet of paper and manually work through how axes should change. We can avoid this slow and often frustrating process by supporting optional axes names for tensors and having the most common ops annotate their outputs:

```python
image = tf.name_axes(image, ('B', 'H', 'W', 'C'))  # assert len(image.shape) == 4
print(image)  # <tf.Tensor dtype=uint8 shape={B: 500, H: 64, W: 64, C: 3}>

grid = tf.repeat(tf.repeat(image, 4, axis='H'), 4, axis='W')
print(grid)  # <tf.Tensor dtype=uint8 shape={B: 500, H: 256, W: 256, C: 3}>

gray = tf.reduce_mean(grid, axis='C')
print(gray)  # <tf.Tensor dtype=uint8 shape={B: 500, H: 256, W: 256}>

flipped = tf.transpose(gray[0], ('W', 'H'))
print(flipped)  # <tf.Tensor dtype=uint8 shape={W: 256, H: 256}>
```

A function starting with the line `tf.name_axes(...)` for its input tensors will immediately be easier to read. Moreover, developers could place `assert tensor.shape.names == ('B', 'F')` statements throughout their code so it becomes immediately clear from the written code how shapes are changing.

An advanced version of named axes would be to keep track of combined axes:

<details>

```python
video = tf.name_axes(image, ('B', 'T', 'H', 'W', 'C'))
print(video)  # <tf.Tensor dtype=uint8 shape={B: 500, T: 100, H: 64, W: 64, C: 3}>

frames = tf.reshape(frames, ('B*T', 'H', 'W', C'))
print(frames)  # <tf.Tensor dtype=uint8 shape={B*T: 50000, H: 64, W: 64, C: 3}>

logits = model(frames)
print(logits)  # <tf.Tensor dtype=float32 shape={B*T: 50000, 10}>

logits = tf.reshape(logits, ('B', 'T', 10))
print(logits)  # <tf.Tensor dtype=float32 shape={B: 500, T: 100, 10}>

logits = tf.name_axes(logits, ('B, 'T', 'K'))
print(logits)  # <tf.Tensor dtype=float32 shape={B: 500, T: 100, K: 10}>
```

</details>

**Will this change the current api? How?**

This would extend the current API in a fully backward compatible way. Common low-level TensorFlow operations (e.g. reduce_sum, concat, slice, tile) have to check for input tensors with named axes to annotate their outputs correctly.

**Who will benefit with this feature?**

Named axes have the potential to increase productivity of both researchers and engineers. Especially when developing large TensorFlow applications, named axes can help keep the overview. For product development with multiple developers, it can increase readability.

**Any Other info.**

JAX supports axes names. PyTorch 1.3 added experimental support for [named dimensions](https://pytorch.org/docs/stable/named_tensor.html).

**Comment by Jaskaran170599** [comment]
I would like to contribute on this feature, if any mentor could guide me.

**Comment by randomstuff** [comment]
What would be the semantic of `x + y` (for example) ?

```python
a = tf.name_axes(a, ('x', 'y'))
b = tf.name_axes(b, ('y', 'x'))
a + b
```

AFAIU, in PyTorch, this currently yields a check that `x.names[i]` is consistent with `y.names[i]` for all `i`: it ensure that you don't mix up dimensions but you still have to keep track of them:

I'd tend to think that automatically reordering dimensions would be more helpful (but this would probably preclude mixing named and unamed tensors and having partially names tensors).

A workaround would be to define helper functions such as:

```python
def plus_by_name(x, y):
    return x + y.align_as(x)
```

**Comment by hosford42** [comment]
[Here](http://nlp.seas.harvard.edu/NamedTensor)'s a relevant article with some ideas on how some higher-level interfaces could be improved using named dimensions. It's written using Torch as an example, but but many of the points apply to any tensor library. Really looking forward to this feature being implemented!

**Comment by danijar** [comment]
Many good ideas in the article you linked!

**Comment by danijar** [comment]
Relevant library that's compatible with TF and other frameworks: https://github.com/arogozhnikov/einops

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by hosford42** [comment]
@Venkat6871 This is a feature request. There is no debug information, and the code is unlikely to become out of date unless the feature has already been implemented with a different interface. Please do not auto-close. I would really like to see this feature.

**Comment by Venkat6871** [comment]
Hi @danijar ,
Thank you so much for filing this feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!
